xdsl/ir.py

```diff
@@ -1000,25 +1118,40 @@ def insert_op(self,
         Insert one or multiple operations at a given index in the block.
         The operations should not be attached to another block.
         """
-        if index < 0 or index > len(self.ops):
-            raise ValueError(
-                f"Can't insert operation in index {index} in a block with "
-                f"{len(self.ops)} operations.")
+
+        # if index < 0 or index > len(self.ops):
+        #     raise ValueError(
+        #         f"Can't insert operation in index {index} in a block with "
+        #         f"{len(self.ops)} operations.")
         if not isinstance(ops, list):
             ops = [ops]
-        if name:
-            for curr_op in ops:
-                for res in curr_op.results:
-                    res.name = name
-        for op in ops:
-            self._attach_op(op)
-        self.ops = self.ops[:index] + ops + self.ops[index:]
+
+        first_op = self.first_op
+
+        if first_op is None:
+            if index:
+                raise ValueError(
+                    f"Can't insert operation in index {index} in a block with "
+                    f"no operations.")
+
+            self.add_ops(ops)
+            return
+
+        if index == 0:
+            # set new head, then add after
+            prev_op = ops[0]
+            ops = ops[1:]
+            self.insert_op_before(prev_op, first_op, name)
+        else:
+            prev_op = self.op_at_index(index - 1)
+
+        self.insert_ops_after(ops, prev_op, name)
```

Comment:
Can you add a comment here explaining what this does? It seems a bit weird at first glance, and I'm having difficulty understanding all the interactions exactly.